ehr_billing/src/org/labkey/ehr_billing/pipeline/BillingTask.java

```diff
@@ -452,15 +452,226 @@ private Collection<Map<String, Object>> getRowList(BillingPipelineJobProcess pro
+
+                //iterate through procedure query results
+                for (Map<String, Object> procedureRow : procedureRows)
+                {
+                    // for each row, associate chargeId
+                    procedureRow.put("chargeId", chargeId);
+
+                    // add additional charge info
+                    ChargeInfo ci = getChargeInfo(chargeId, chargeInfoArrayList, (Date) procedureRow.get("date"));
+                    procedureRow.put("item", ci.getItem());
+                    procedureRow.put("category", ci.getCategory());
+
+                    // get cost
+                    Double unitCost = ci.getUnitCost();
+                    procedureRow.put(processingService.getUnitCostColName(), unitCost);
+
+                    // total cost
+                    Double totalCost = unitCost * (Double) procedureRow.get("quantity");
+                    procedureRow.put(processingService.getTotalCostColName(), totalCost);
+
+                    // calculate total cost with additional/other rate (ex. tier rate for WNPRC)
+                    Double otherRate = (Double) procedureRow.get("otherRate");
+                    Double unitCostWithOtherRate;
+                    Double totalCostWithOtherRate;
+                    if (null != otherRate &&
+                            null != processingService.getAdditionalUnitCostColName() &&
+                            null != processingService.getAdditionalTotalCostColName())
+                    {
+                        unitCostWithOtherRate = unitCost + (unitCost * otherRate);
+                        procedureRow.put(processingService.getAdditionalUnitCostColName(), unitCostWithOtherRate);
 
+                        totalCostWithOtherRate = unitCostWithOtherRate * (Double) procedureRow.get("quantity");
+                        procedureRow.put(processingService.getAdditionalTotalCostColName(), totalCostWithOtherRate);
+                    }
+                }
+                writeToInvoicedItems(process, procedureRows, getSupport());
+            }
+        }
+        else if (process.getQueryName() != null && !process.getQueryToInvoiceItemColMap().isEmpty())
+        {
             Collection<Map<String, Object>> rows = getRowList(process, billingRunContainer);
             getJob().getLogger().info(rows.size() + " rows found");
 
             writeToInvoicedItems(process, rows, getSupport());
-            getJob().getLogger().info("Finished Caching " + process.getLabel());
+        }
+        getJob().getLogger().info("Finished Caching " + process.getLabel());
+    }
+
+    private ChargeInfo getChargeInfo(int chargeId, ArrayList<ChargeInfo> chargeInfoArrayList, Object date)
+    {
+        Date chargeDate = (Date) date;
+        List<ChargeInfo> chargeInfoList = chargeInfoArrayList.stream().filter(ci -> ci.getChargeId() == chargeId).toList();
+
+        if (chargeInfoList.size() == 1)
```

Comment:
Similar feedback to the other PR - no need for these `if` checks. Just the `for` loop is enough

Reply:
yup, now updated